src/core/rest-types.ts

```diff
@@ -0,0 +1,71 @@
+import { ChatMessageAction } from './events.js';
+import { DefaultMessage, emptyMessageReactions, Message, MessageHeaders, MessageMetadata } from './message.js';
+
+// RestClientIdList represents a list of client IDs with aggregation data
+export interface RestClientIdList {
+  total: number;
+  clientIds: string[];
+}
+
+// RestClientIdCounts represents client ID counts with aggregation data
+export interface RestClientIdCounts {
+  total: number;
+  clientIds: Record<string, number>;
+  totalUnidentified: number;
+}
+
+// RestOperation represents an operation performed on a chat message
+export interface RestOperation {
+  clientId?: string;
+  description?: string;
+  metadata?: Record<string, string>;
+}
+
+// ChatMessageReactions represents reactions on a chat message
+export interface RestChatMessageReactions {
+  unique?: Record<string, RestClientIdList>;
+  distinct?: Record<string, RestClientIdList>;
+  multiple?: Record<string, RestClientIdCounts>;
+}
+
+// RestMessage represents a message in V3 of the REST API.
+export interface RestMessage {
+  serial: string;
+  version: string;
+  roomId: string;
+  text: string;
+  clientId: string;
+  action: string;
+  metadata: Record<string, unknown>;
+  headers: Record<string, string>;
+  createdAt: number;
+  timestamp: number;
+  operation?: RestOperation;
+  reactions?: RestChatMessageReactions;
+}
+
+/**
+ * Converts a message object from its REST representation to the standard message object in the SDK.
+ *
+ * @param message The message to convert from REST.
+ * @returns The converted message.
+ */
+export const messageFromRest = (message: RestMessage): Message => {
+  const metadata = message.metadata as MessageMetadata | undefined;
+  const headers = message.headers as MessageHeaders | undefined;
+
+  const reactions = {
+    ...emptyMessageReactions(),
+    ...message.reactions,
+  };
+
+  return new DefaultMessage({
+    ...message,
+    action: message.action as ChatMessageAction,
```

Comment:
same goes here - if it's already gonna be a `ChatMessageAction` then we should express that in the `RestMessage` type

Reply:
I've made the types better defined, and added a default in [1e0afc2](https://github.com/ably/ably-chat-js/pull/571/commits/1e0afc2db39681f711c7fc44d76911d267e228a7)